api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+			"sendfil": logging.LevelDebug,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.MongoUri))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.MongoDbName)
+	col := db.Collection(collectionName)
+	if _, err := col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{primitive.E{Key: "from", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "to", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "amount_nano_fil", Value: 1}},
+		},
+		// MongoDB automatically creates a multikey index if any indexed field is an array;
+		// you do not need to explicitly specify the multikey type.
+		// https://docs.mongodb.com/manual/core/index-multikey/
+		{
+			Keys:    bson.D{primitive.E{Key: "message_cids.cid", Value: 1}},
+			Options: options.Index().SetUnique(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "message_state", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "waiting", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "updated_at", Value: 1}},
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("creating collection indexes: %v", err)
+	}
+
+	s := &Service{
+		clientBuilder: config.ClientBuilder,
+		col:           col,
+		waiting:       make(map[primitive.ObjectID]chan waitResult),
+		config:        config,
+		ticker:        time.NewTicker(config.RetryWaitFrequency),
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterSendFilServiceServer(s.server, s)
+		if err := s.server.Serve(config.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	if err := s.waitAllQualifying(ctx, true); err != nil {
+		return nil, fmt.Errorf("calling waitAllQualifying: %v", err)
+	}
+
+	s.bindTicker(ctx)
+
+	return s, nil
+}
+
+func (s *Service) waitAllQualifying(ctx context.Context, isInitialRun bool) error {
+	filter := bson.M{"$and": bson.A{
+		bson.M{"message_state": bson.M{"$ne": pb.MessageState_MESSAGE_STATE_ACTIVE}},
+		bson.M{"message_state": bson.M{"$ne": pb.MessageState_MESSAGE_STATE_FAILED}},
+	}}
+	if !isInitialRun {
+		filter["waiting"] = false
+	}
+	cursor, err := s.col.Find(ctx, filter)
+	if err != nil {
+		return status.Errorf(codes.Internal, "querying txns: %v", err)
+	}
+	defer cursor.Close(ctx)
+	var txns []txn
+	err = cursor.All(ctx, &txns)
+	if err != nil {
+		return status.Errorf(codes.Internal, "decoding txns query results: %v", err)
+	}
+	log.Infof("found %v txns to initiate waiting on", len(txns))
+	for _, txn := range txns {
+		s.wait(txn)
+	}
+	return nil
+}
+
+func (s *Service) bindTicker(ctx context.Context) {
+	go func() {
+		for {
+			<-s.ticker.C
+			if err := s.waitAllQualifying(ctx, false); err != nil {
+				log.Errorf("waitAllQualifying from ticker: %v", err)
+			}
+		}
+	}()
```

Comment:
Pretty simple here, whenever the ticker ticks, call `waitAllQualifying`, passing `false` since it's not the initial run.